### PR TITLE
Improve defaults generation with logging, stats, and token validation

### LINKED ISSUE #122602: Provide list of used Material Design tokens
To aid in determining coverage of design tokens usage, we should provide at least a list of used tokens (e.g. `md.comp.switch.color`).

To do so, we either have to enforce that all tokens are used by their full string (i.e. no concatenation), or are logged when they are resolved.

### PR DESCRIPTION
**Pull request by guidezpl**
## Description

This improves defaults generation with logging, stats, and token validation. 

This PR includes these changes:
* introduce `TokenLogger`, with a verbose mode
  * prints versions and tokens usage to the console
  * outputs `generated/used_tokens.csv`, a list of all used tokens, for use by Google
* find token files in `data` automatically
* hide tokens `Map`
  * tokens can be obtained using existing resolvers (e.g. `color`, `shape`), or directly through `getToken`.
  * tokens can be checked for existence with `tokenAvailable`
* remove version from template, since the tokens are aggregated and multiple versions are possible (as is the case currently), it does not make sense to attribute a single version
* improve documentation

## Related Issues
 - Fixes https://github.com/flutter/flutter/issues/122602

## Tests
 - Added tests for `TokenLogger`
 - Regenerated tokens, no-op except version removal

## Future work
A future PR should replace or remove the following invalid tokens usages

image

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: dev/tools/gen_defaults/generated/used_tokens.csv [hunk 1]
```diff
@@ -0,0 +1,926 @@
...
+md.sys.typescale.title-large.tracking,
+md.sys.typescale.title-large.weight,
+md.sys.typescale.title-medium.line-height,
+md.sys.typescale.title-medium.size,
+md.sys.typescale.title-medium.tracking,
+md.sys.typescale.title-medium.weight,
+md.sys.typescale.title-small.line-height,
+md.sys.typescale.title-small.size,
+md.sys.typescale.title-small.tracking,
+md.sys.typescale.title-small.weight
```
Comment (QuncCccccc) [inline dev/tools/gen_defaults/generated/used_tokens.csv:926]: nit: I saw a minus sign here, should here be a new line?
Screenshot 2023-06-07 at 11 00 41 PM

File: dev/tools/gen_defaults/lib/color_scheme_template.dart [hunk 2]
```diff
@@ -3,86 +3,97 @@
 // found in the LICENSE file.
 
 import 'template.dart';
+import 'token_logger.dart';
 
 class ColorSchemeTemplate extends TokenTemplate {
-  ColorSchemeTemplate(super.blockName, super.fileName, super.tokens);
+  ColorSchemeTemplate(this._colorTokensLight, this._colorTokensDark, super.blockName, super.fileName, super.tokens);
```
Comment (QuncCccccc) [inline dev/tools/gen_defaults/lib/color_scheme_template.dart:9]: Just for my education, why do we use private properties in constructor?

File: dev/tools/gen_defaults/lib/template.dart [hunk 3]
```diff
@@ -31,6 +34,15 @@ abstract class TokenTemplate {
   /// Defaults to 'Theme.of(context).textTheme.'
   final String textThemePrefix;
 
+  /// Check if a token is available.
+  bool tokenAvailable(String tokenName) => _tokens.containsKey(tokenName);
+
+  /// Resolve a token while logging its usage.
+  dynamic getToken(dynamic tokenName) {
```
Comment (QuncCccccc) [inline dev/tools/gen_defaults/lib/template.dart:41]: Should these two `dynamic` just be `String`?

**Review by QuncCccccc — COMMENTED** [review]
This looks great! Just left some questions below:)

File: dev/tools/gen_defaults/lib/color_scheme_template.dart (hunk 2 above)
Comment (guidezpl) [reply dev/tools/gen_defaults/lib/color_scheme_template.dart:9]: It's no different than the following

ColorSchemeTemplate(
    Map<String, dynamic> colorTokensLight,
    ...
) : _colorTokensLight = colorTokensLight;

File: dev/tools/gen_defaults/generated/used_tokens.csv (hunk 1 above)
Comment (guidezpl) [reply dev/tools/gen_defaults/generated/used_tokens.csv:926]: Fixed

File: dev/tools/gen_defaults/lib/template.dart (hunk 3 above)
Comment (guidezpl) [reply dev/tools/gen_defaults/lib/template.dart:41]: Yes for `tokenName`, no for the return type, which can be other types (e.g. `double`)

**Review by QuncCccccc — APPROVED** [review]
LGTM:)!!

**Pull request merged**